x/leverage/keeper/token.go

```diff
@@ -5,6 +5,36 @@ import (
 	"github.com/umee-network/umee/v4/x/leverage/types"
 )
 
+// CleanTokenRegistry deletes all blacklisted tokens in the leverage registry
+// whose uToken supplies are zero. Called automatically on registry update.
+func (k Keeper) CleanTokenRegistry(ctx sdk.Context) error {
+	tokens := k.GetAllRegisteredTokens(ctx)
+	for _, t := range tokens {
+		if t.Blacklist {
+			uDenom := types.ToUTokenDenom(t.BaseDenom)
+			uSupply := k.GetUTokenSupply(ctx, uDenom)
+			if uSupply.IsZero() {
+				err := k.deleteTokenSettings(ctx, t)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
```

Comment:
If the uToken supply of a token is zero, then:
- There are no tokens borrowed out or stored in the module account (so base tokens are okay)
- No uTokens are held by users or the module (no supply or collateral exists)

And if it's also blacklisted
- Supply and borrow messages are already disabled, so there's no risk of a message coming through that would interfere with the removal. Since this check is happening at the instant the proposal takes effect, I don't think it would be possible to sneak a message in between either.

This approach avoids the problem where we remove a token with existing uTokens or borrows and leverage no longer knows how to handle it (would lead to all kinds of problems).

Reply:
Yes, but why do we want to remove the token at all?
Anyways, going to pre-approve. Please add an info about token removal in the leverage/README

Reply:
Added a note. Token removal is only safe for things that have never been supplied before they were blacklisted - in practice, it will apply if we get an IBC denom wrong on mainnet, or list something we never use on a testnet.